mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoSafeUpdateProtocol.java

```diff
@@ -0,0 +1,349 @@
+                }
+                LOGGER.debug("writeResult={}",writeResult);
+                if(idsInBatch.size()==writeResult.getMatchedCount()) {
+                    LOGGER.debug("Successful update");
+                } else {
+                    LOGGER.debug("notUpdated={}",idsInBatch.size()-writeResult.getMatchedCount());
+                    findConcurrentModifications(results);
+                }
+            } catch (BulkWriteException e) {
+                List<BulkWriteError> writeErrors=e.getWriteErrors();
+                if(writeErrors!=null) {
+                    for(BulkWriteError we:writeErrors) {
+                        if (we.getCode() == 11000 || we.getCode() == 11001) {
+                            results.put(we.getIndex(),
+                                        Error.get("update", MongoCrudConstants.ERR_DUPLICATE, we.getMessage()));
+                        } else {
+                            results.put(we.getIndex(),
+                                        Error.get("update", MongoCrudConstants.ERR_SAVE_ERROR, we.getMessage()));
+                        }
+                    }
+                }
+                findConcurrentModifications(results);
+            }
+        }
+        reset();
+        return results;
+    }
+
+    /**
+     * This executes a query to find out documents with concurrent modification errors
+     */
+    protected void findConcurrentModifications(Map<Integer,Error> results) {
+        List<Object> updatedIds=new ArrayList<>(idsInBatch.size());
+        // Collect all ids without errors
+        int index=0;
+        for(Object id:idsInBatch) {
+            if(!results.containsKey(index))
+                updatedIds.add(id);
+            index++;
+        }
+        LOGGER.debug("checking for concurrent modifications:{}",updatedIds);
+        if(!updatedIds.isEmpty()) {
+            // documents with the given _ids and whose docver contains our docVer are the ones we managed to update
+            // others are failures
+            BasicDBObject query=new BasicDBObject(DOCVER_FLD,docVer);
+            query.append("_id",new BasicDBObject("$in",updatedIds));
+            DBCursor cursor=null;
+            try {
+                Set<Object> successfulIds=new HashSet<>(updatedIds.size());
+                cursor=collection.find(query,new BasicDBObject("_id",1));
+                while(cursor.hasNext()) {
+                    DBObject obj=cursor.next();
+                    Object id=obj.get("_id");
+                    successfulIds.add(id);
+                }
+                index=0;
+                for(Object id:idsInBatch) {
+                    if(!results.containsKey(index)) { // No other errors for this id
+                        if(!successfulIds.contains(id)) {
+                            // concurrency errors for this id
+                            results.put(index,Error.get("update",MongoCrudConstants.ERR_CONCURRENT_UPDATE,id.toString()));
```

Comment:
Per above comment, is worth checking here if we may have not seen the doc because the version is too old?

Not sure what you would be able to do about it other than log.